src/cpp/src/sampling/sampler.cpp

```diff
@@ -1017,13 +1019,11 @@ LogitProcessor& Sampler::get_logit_processor(uint64_t request_id) {
 
 
 void Sampler::create_logit_processor(uint64_t request_id, const GenerationConfig& sampling_params, const TokenIds& prompt) {
-    if (!m_structured_output_controller) {
-        // We don't have vocab size (actually logits size) and also we don't have access to the logits.
-        // vocab size will be taken from the tokenizer during LogitProcessor initialization.
-        m_structured_output_controller = std::make_shared<StructuredOutputController>(m_tokenizer);
+    std::shared_ptr<StructuredOutputController> structured_output_controller = nullptr;
+    if (m_tokenizer.m_pimpl != nullptr) {
+        structured_output_controller = m_tokenizer.m_pimpl->get_structured_output_controller();
```

Comment:
I think there's no way to construct an instance of `Tokenizer` with null `m_pimpl`. If you want to verify that, it should be an error to avoid misleading that it's a valid situation if `m_pimpl` is null